### PR TITLE
OSDOCS-3429: OSDK-2129 Bundle validations for Operators

### PR DESCRIPTION
**Pull request by michaelryanpeter**
Version(s): 4.11

Issue: [OSDOCS-3429](https://issues.redhat.com//browse/OSDOCS-3429)

Link to docs preview (VPN required): 
[Validating Operator bundles](http://file.rdu.redhat.com/mipeter/osdocs-3429--osdk-2129_bundle-validations-good-practices/operators/operator_sdk/osdk-bundle-validate.html)

Additional information:

### CODE REVIEW

File: modules/osdk-bundle-validate-about.adoc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+----
+./bundle
+  ├── manifests
+  │   ├── cache.my.domain_memcacheds.yaml
+  │   └── memcached-operator.clusterserviceversion.yaml
+  └── metadata
+      └── annotations.yaml
+----
+
+.`bundle validate` command syntax
```
Comment (sheriff-rh) [inline modules/osdk-bundle-validate-about.adoc:24]: I feel like this could be on line 10, beneath the `bundle validate` description. It's a little awkward beneath the bundle structure layout, IMO.

File: modules/osdk-bundle-validate-tests.adoc [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Module included in the following assemblies:
+//
+// * operators/operator_sdk/osdk-bundle-validate.adoc
+
+:_content-type: REFERENCE
+[id="osdk-bundle-validate-tests_{context}"]
+= Built-in bundle validate tests
+
+The Operator SDK ships with pre-defined `bundle validate` tests arranged into suites. If you run the `bundle validate` command without specifying a test, the default test runs. The default test verifies a bundle adheres to the specifications outlined in the link:https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md["Operator Bundle Overview"] in the upstream Operator Registry repository.
```
Comment (sheriff-rh) [inline modules/osdk-bundle-validate-tests.adoc:9]: I don't think quotation marks are necessary around Operator Bundle Overview.

File: modules/osdk-bundle-validate-about.adoc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+----
+$ operator-sdk bundle validate <bundle_dir_or_image> <flag>
+----
+
+The command completes with an exit code of 0 if it does not detect validation errors. The command completes with an exit code of 1 if it detects validation errors. The command only fails on errors. Bundles with only warnings and no errors can still pass with an exit code of 0.
+
+.Example output
+[source,terminal]
+----
+some output
```
Comment (sheriff-rh) [inline modules/osdk-bundle-validate-about.adoc:35]: I believe you're just using "some output" as a placeholder for the moment, right? No worries if so :+1:

**Review by sheriff-rh — COMMENTED** [review]
Looking good so far! Some small nits.

File: modules/osdk-bundle-validate-about.adoc (hunk 3 above)
Comment (michaelryanpeter) [reply modules/osdk-bundle-validate-about.adoc:35]: It is a placeholder for now until I have some example output.

File: modules/osdk-bundle-validate-about.adoc [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+:_content-type: CONCEPT
+[id="osdk-about-bundle-validate_{context}"]
+= About the bundle validate command
+
+While the Operator SDK `scorecard` command can run tests on your Operator based on a configuration file and test images, the `bundle validate` subcommand can validate local bundle directories and remote bundle images for content and structure.
+
+.`bundle validate` command syntax
+[source,terminal]
+----
+$ operator-sdk bundle validate <bundle_dir_or_image> <flag>
```
Comment (jmrodri) [inline modules/osdk-bundle-validate-about.adoc:14]: There are more than one flag that can be put in
```suggestion
$ operator-sdk bundle validate <bundle_dir_or_image> <flags>
```

File: modules/osdk-bundle-validate-run.adoc [hunk 5]
```diff
@@ -0,0 +1,62 @@
+// Module included in the following assemblies:
+//
+// * operators/operator_sdk/osdk-bundle-validate.adoc
+
+:_content-type: PROCEDURE
+[id="osdk-bundle-validate-run_{context}"]
+= Running the bundle validate command
+
+The default test runs every time you enter the `bundle validate` command. If you want to run optional tests using the `--select-optional` flag, those test run in addition to the default test.
```
Comment (jmrodri) [inline modules/osdk-bundle-validate-run.adoc:9]:
```suggestion
The default test runs every time you enter the `bundle validate` command. If you want to run optional tests using the `--select-optional` flag, those tests run in addition to the default test.
```

File: modules/osdk-bundle-validate-run.adoc [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
++
+where:
+
+<bundle_registry>:: Specifies the registry where the bundle is hosted, such as `quay.io/example`.
+<bundle_image_name>:: Specifies the name of the bundle image, such as `memcached-operator`.
+<tag>:: Specifies the tag of the bundle image, such as `v1.22.0`.
++
+[NOTE]
+====
+If you want to test an Operator bundle image, you must host your image in a remote registry. The Operator SDK pulls the image and builds it locally before running tests. The `bundle validate` command does not support testing local bundle images.
```
Comment (jmrodri) [inline modules/osdk-bundle-validate-run.adoc:40]: Is this referring to the case where I can build an image and it is in my docker image list but not pushed? If so, then this is correct. When I think of local bundle images I think of the bundle directory :) but I think that's because I know that you have to push the image.

File: modules/osdk-bundle-validate-run.adoc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+----
+$ operator-sdk bundle validate \
+  <bundle_dir_or_image> \
+  --select-optional <test_label>
+----
++
+where:
+
+<bundle_dir_or_image>:: Specifies the local bundle directory or remote bundle image such as `~/projects/memcached` or `quay.io/example/memcached-operator:v1.22`.
+<test_label>:: Specifies the name of the test you want to run, such as `name=good-practices`.
```
Comment (jmrodri) [inline modules/osdk-bundle-validate-run.adoc:55]: We call them tests because validator isn't word right?

File: modules/osdk-bundle-validate-tests.adoc [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+//
+// * operators/operator_sdk/osdk-bundle-validate.adoc
+
+:_content-type: REFERENCE
+[id="osdk-bundle-validate-tests_{context}"]
+= Built-in bundle validate tests
+
+The Operator SDK ships with pre-defined `bundle validate` tests arranged into suites. If you run the `bundle validate` command without specifying a test, the default test runs. The default test verifies a bundle adheres to the specifications in the upstream link:https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md[Operator Bundle Overview].
+
+You can run your bundle against additional tests, such as the requirements necessary to publish your Operator on the OperatorHub or to check for deprecated Kubernetes APIs. Any additional tests run along with the default test.
```
Comment (jmrodri) [inline modules/osdk-bundle-validate-tests.adoc:11]: The "Any additional tests run along with the default test." feels awkward like there should be more or rephrased.

Comment (jmrodri) [reply modules/osdk-bundle-validate-tests.adoc:11]: Maybe I'm parsing it wrong. So when I read it, I see the beginning of a sentence that says additional tests run along with the default test WILL DO SOMETHING ELSE. 

but I think what you are trying to say that "additional tests" always run in addition to the default test. 

If the latter is what you mean, then I think it is phrased correctly, I just parsed it wrong.

File: modules/osdk-bundle-validate-tests.adoc [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+----
+
+[id="osdk-bundle-validate-additional-tests_{context}"]
+.Addtional `bundle validate` tests
+[cols="3,7,3",options="header"]
+|===
+|Name |Description |Label
+
+|Operator Framework
+|This test validates a bundle against the entire suite of validators provided by the Operator Framework.
```
Comment (jmrodri) [inline modules/osdk-bundle-validate-tests.adoc:27]: Surprised you actually used the term _validators_ :)

**Review by jmrodri — APPROVED** [review]
/lgtm

File: modules/osdk-bundle-validate-tests.adoc (hunk 8 above)
Comment (michaelryanpeter) [reply modules/osdk-bundle-validate-tests.adoc:11]: I have tried to rephrase it to make it clearer. If you need to interpret and think abou what I am trying to say then a customer is likely to have a hard time too.

**Comment by michaelryanpeter** [comment]
@emmajiafan Would you PTAL?

File: modules/osdk-bundle-validate-tests.adoc [hunk 10]
```diff
@@ -0,0 +1,37 @@
+// Module included in the following assemblies:
+//
+// * operators/operator_sdk/osdk-bundle-validate.adoc
+
+:_content-type: REFERENCE
+[id="osdk-bundle-validate-tests_{context}"]
+= Built-in bundle validate tests
+
+The Operator SDK ships with pre-defined validators arranged into suites. If you run the `bundle validate` command without specifying a validator, the default test runs. The default test verifies a bundle adheres to the specifications defined by the Operator Framework community. For more information, see "Bundle format".
```
[the remark this answers is not in the document]
Comment (michaelryanpeter) [reply modules/osdk-bundle-validate-tests.adoc:9]: Unfortunately, our docs tooling does not let us make cross references (internal links) in modules, only at the docs assembly level. The link is in the [Additional resources section here.](https://github.com/openshift/openshift-docs/pull/47613/files#diff-6bff3e28092b57ec4866c43bd7489024c8818dde396200ebb72c1dac3ac840bcR17)

[the remark this answers is not in the document]
Comment (michaelryanpeter) [reply modules/osdk-bundle-validate-tests.adoc:9]: @jmrodri

File: modules/osdk-bundle-validate-about.adoc [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+----
+$ operator-sdk bundle validate <bundle_dir_or_image> <flags>
+----
+
+[NOTE]
+====
+The `bundle validate` command runs automatically when you build your bundle using the the `make bundle` command. 
+====
+
+Bundle images are pulled from a remote registry and built locally before they are validated. Local bundle directories must contain Operator metadata and manifests. The bundle metadata and manifests should have a structure similar to the following bundle layout:
```
Comment (tmalove) [inline modules/osdk-bundle-validate-about.adoc:22]:
```suggestion
Bundle images are pulled from a remote registry and build locally before they are validated. Local bundle directories must contain Operator metadata and manifests. The bundle metadata and manifests should have a structure similar to the following bundle layout:
```

File: modules/osdk-bundle-validate-run.adoc [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+[source,terminal]
+----
+$ operator-sdk bundle validate \
+  <bundle_dir_or_image> \
+  --select-optional <test_label>
+----
++
+where:
+
+<bundle_dir_or_image>:: Specifies the local bundle directory or remote bundle image such as `~/projects/memcached` or `quay.io/example/memcached-operator:v1.22`.
```
Comment (tmalove) [inline modules/osdk-bundle-validate-run.adoc:54]:
```suggestion
<bundle_dir_or_image>:: Specifies the local bundle directory or remote bundle image, such as `~/projects/memcached` or `quay.io/example/memcached-operator:v1.22`.
```

File: modules/osdk-bundle-validate-tests.adoc (hunk 10 above)
Comment (tmalove) [inline modules/osdk-bundle-validate-tests.adoc:9]:
```suggestion
The Operator SDK ships with pre-defined validators arranged into suites. If you run the `bundle validate` command without specifying a validator, the default test runs. The default test verifies that a bundle adheres to the specifications defined by the Operator Framework community. For more information, see "Bundle format".
```

File: modules/osdk-bundle-validate-tests.adoc [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+|Operator Framework
+|This validator tests an Operator bundle against the entire suite of validators provided by the Operator Framework.
+|`suite=operatorframework`
+
+|OperatorHub
+|This validator tests an Operator bundle for compatibility with OperatorHub.
+|`name=operatorhub`
+
+|Good Practices
+|This validator tests an Operator bundle complies with good practices as defined by the Operator Framework. It checks for issues such as an empty CRD description or unsupported Operator Lifecycle Manager (OLM) resources.
```
Comment (tmalove) [inline modules/osdk-bundle-validate-tests.adoc:35]:
```suggestion
|This validator tests an Operator bundle complies with good practices as defined by the Operator Framework. It checks for issues, such as an empty CRD description or unsupported Operator Lifecycle Manager (OLM) resources. 
```

**Review by kalexand-rh — COMMENTED** [review]
This is looking good!

**Review by jldohmann — APPROVED** [review]
/lgtm
